waitress/app/utils.py

```diff
@@ -128,98 +130,82 @@ def difference(cls, repo_users, db_users):
-                    # use slack default image
-                    'photo': item['profile'].get('image_original', item['profile'].get('image_192')),
-                    'firstname': firstname.title(),
-                    'lastname': lastname.title()
-                }
-
-            return user_dict, not_user_list
-
-        info, not_user = normalize(info)
-
-        difference = cls.get_real_users(difference[:], not_user)
-
-        if len(difference) == 0:
-            return "Users list not changed"
+                user_slack_id = item['id']
+
+                if is_user_invalid(item):
+                    invalid_users.append(user_slack_id)
+                else:
+                    firstname = item.get('profile').get('first_name', '')
+                    lastname = item.get('profile').get('last_name', '')
+                    valid_users[user_slack_id] = {
+                        'slack_id': user_slack_id,
+                        'email': item['profile']['email'],
+                        # use slack default image
+                        'photo': item['profile'].get('image_original', item['profile'].get('image_192')),
+                        'firstname': firstname.title(),
+                        'lastname': lastname.title()
+                    }
+
+            return valid_users, invalid_users
+
+        valid_users, invalid_users = normalize(info)
+
+        valid_slack_users = [user for user in difference if user not in invalid_users]
+
+        if len(valid_slack_users) == 0:
+            return "No new user found on slack."
 
         if trim:
-            return cls.trim_away(not_user)
+            return cls.trim_away(invalid_users)
 
         try:
-            for user in difference:
-                if user in info:
-                    with transaction.atomic():
-                        SlackUser.create(info[user])
-            return "Users updated successfully"
+            for user_slack_id in valid_slack_users:
+                current_user = valid_users[user_slack_id]
+                with transaction.atomic():
+                    SlackUser.create(current_user)
+            return "Users updated successfully."
         except Exception as e:
-            return "Users couldn't be updated successfully - %s" % e.message
-
-    @staticmethod
-    def get_real_users(difference, not_user):
-        """Get the users that are legit."""
-        return [item for item in difference if item not in not_user]
+            return f"Users couldn't be updated successfully - {e.args[0]}"
 
     @classmethod
-    def trim_away(cls, not_user):
+    def trim_away(cls, invalid_users):
         """Trims away users for the system that have been deleted off Slack."""
         trim_user = []
         for user in cls.user_queryset:
-            if user.slack_id in not_user:
-                trim_user.append("%s %s" % (user.firstname, user.lastname))
+            if user.slack_id in invalid_users:
```

Comment:
use comprehension for the trim_user and query for the delete all users that are affected at a go instead of multiple queries